modules/optimization/doc/content/source/reporters/VectorDotProduct.md

```diff
@@ -0,0 +1,13 @@
+# VectorDotProduct
```

Comment:
actually StdVectorRealReporterParsedOperation and you pass a parsed expression that is applied to every component

Reply:
this is good idea, I have to think about it. I could easily do the stuff in https://mooseframework.inl.gov/source/postprocessors/VectorPostprocessorReductionValue.html
but I don't think this is how you'd want to generalize reporter operations that contain Real values.   I just realized I need to add another reporter that will sum two scalars.  So now there is a row sum for a vector of vectors, vector dot product, vector sum into a scalar and scalar sum.

Reply:
@loganharbour do you think we handle the type of the reporter dynamically?
As in we pass
"input_reporter_type = Vector"
"input_data_type = Real"
"output_reporter_type = Scalar"
"output_data_type = Real"

Reply:
I'm pretty sure that we'll be required to make the type an input option

Reply:
I agree with the usefulness of @GiudGiud and @loganharbour comments about a more general reporter with operations object but I'm not the right person to develop something like that.  In the mean time, I can get these reporters and PR merged and replace them later when a better option becomes available. 

These reporters that operate on their data aren't useful for standard simulations because they are just doing a post-processing step that could easily be done in python since all of the reporter data is going to be written to json anyway.  So maybe it is not worth the effort to develop a more general reporter with operations object.  For optimization, I need to process the data and then make a decision using that data within the same simulation, so this can't be done as a postprocessing step.

Reply:
There's plenty of examples of parsed operation objects at this point, it's not as complicated as it used to be. 

I dont think it's sustainable to implement an object for basic operations, and while this was the initial approach for auxkernels and PPs, it was just not the right approach. We cant solve a combinatorial problem (N operations * M input data types * M output data types * L systems (auxkernels, pps, reporters, vpps...)) by addressing every combination individually. I think doing parsing + templating would get us though the N*M*M part of the problem.